File: benchmark_client/scripts/go/Pandemic Spread.go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"time"
)

const (
	susceptible = 0
	infected    = 1
	recovered   = 2
	N           = 10000
	K           = 4
	beta        = 0.1
	lambdaInf   = 0.3
	gamma       = 0.1
	maxDays     = 200
)

func buildSmallWorld(src *rand.Rand) [][]int {
	adj := make([][]int, N)
	for i := range adj {
		adj[i] = nil
	}
	for i := 0; i < N; i++ {
		for j := 1; j <= K/2; j++ {
			right := (i + j) % N
			if src.Float64() < beta {
				nt := src.Intn(N)
				for nt == i {
					nt = src.Intn(N)
				}
				adj[i] = append(adj[i], nt)
				adj[nt] = append(adj[nt], i)
			} else {
				adj[i] = append(adj[i], right)
				adj[right] = append(adj[right], i)
			}
		}
	}
	return adj
}

func epidemicStep(stIn, stOut []int, adj [][]int, src *rand.Rand) {
	for idx := 0; idx < N; idx++ {
		my := stIn[idx]
		if my == recovered {
			stOut[idx] = recovered
		} else if my == infected {
			if src.Float64() < gamma {
				stOut[idx] = recovered
			} else {
				stOut[idx] = infected
			}
		} else {
			gets := false
			for _, nb := range adj[idx] {
				if stIn[nb] != infected {
					continue
				}
				if src.Float64() < lambdaInf {
					gets = true
					break
				}
			}
			if gets {
				stOut[idx] = infected
			} else {
				stOut[idx] = susceptible
			}
		}
	}
}

func main() {
	src := rand.New(rand.NewSource(42))
	adj := buildSmallWorld(src)

	state := make([]int, N)
	nxt := make([]int, N)
	for i := 0; i < 5; i++ {
		state[src.Intn(N)] = infected
	}

	start := time.Now()
	for d := 0; d < maxDays; d++ {
		epidemicStep(state, nxt, adj, src)
		state, nxt = nxt, state
	}
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6

	_ = state[0]
	out, err := json.Marshal(map[string]float64{"execution_time_ms": elapsedMs})
	if err != nil {
		os.Exit(1)
	}
	fmt.Println(string(out))
}

File: benchmark_client/scripts/cpp/Pandemic Spread.cpp
#include <chrono>
#include <cstdio>
#include <random>
#include <vector>

enum { SUSCEPTIBLE = 0, INFECTED = 1, RECOVERED = 2 };

static void build_small_world(int N, int K, float beta, std::mt19937 &gen,
                             std::vector<std::vector<int>> &adj) {
    adj.assign(static_cast<std::size_t>(N), {});
    std::uniform_real_distribution<float> uni(0.0f, 1.0f);
    std::uniform_int_distribution<int> rnode(0, N - 1);
    for (int i = 0; i < N; ++i) {
        for (int j = 1; j <= K / 2; ++j) {
            const int right = (i + j) % N;
            if (uni(gen) < beta) {
                int nt = rnode(gen);
                while (nt == i) {
                    nt = rnode(gen);
                }
                adj[static_cast<std::size_t>(i)].push_back(nt);
                adj[static_cast<std::size_t>(nt)].push_back(i);
            } else {
                adj[static_cast<std::size_t>(i)].push_back(right);
                adj[static_cast<std::size_t>(right)].push_back(i);
            }
        }
    }
}

static void epidemic_step(const std::vector<int> &st_in, std::vector<int> &st_out,
                          const std::vector<std::vector<int>> &adj, int N, float lambda_inf,
                          float gamma, std::mt19937 &gen) {
    std::uniform_real_distribution<float> uni(0.0f, 1.0f);
    for (int idx = 0; idx < N; ++idx) {
        const int my = st_in[static_cast<std::size_t>(idx)];
        if (my == RECOVERED) {
            st_out[static_cast<std::size_t>(idx)] = RECOVERED;
        } else if (my == INFECTED) {
            if (uni(gen) < gamma) {
                st_out[static_cast<std::size_t>(idx)] = RECOVERED;
            } else {
                st_out[static_cast<std::size_t>(idx)] = INFECTED;
            }
        } else {
            bool gets = false;
            for (int nb : adj[static_cast<std::size_t>(idx)]) {
                if (st_in[static_cast<std::size_t>(nb)] != INFECTED) {
                    continue;
                }
                if (uni(gen) < lambda_inf) {
                    gets = true;
                    break;
                }
            }
            st_out[static_cast<std::size_t>(idx)] = gets ? INFECTED : SUSCEPTIBLE;
        }
    }
}

int main() {
    constexpr int N = 10000;
    constexpr int K = 4;
    constexpr float beta = 0.1f;
    constexpr float lambda_inf = 0.3f;
    constexpr float gamma = 0.1f;
    constexpr int max_days = 200;

    std::mt19937 gen(42u);
    std::vector<std::vector<int>> adj;
    build_small_world(N, K, beta, gen, adj);

    std::vector<int> state(static_cast<std::size_t>(N), SUSCEPTIBLE);
    std::uniform_int_distribution<int> rnode(0, N - 1);
    for (int i = 0; i < 5; ++i) {
        state[static_cast<std::size_t>(rnode(gen))] = INFECTED;
    }

    std::vector<int> nxt(static_cast<std::size_t>(N));

    const auto t0 = std::chrono::high_resolution_clock::now();
    for (int d = 0; d < max_days; ++d) {
        epidemic_step(state, nxt, adj, N, lambda_inf, gamma, gen);
        state.swap(nxt);
    }
    const auto t1 = std::chrono::high_resolution_clock::now();
    const double elapsed_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    (void)state[0];
    std::printf("{\"execution_time_ms\": %.6f}\n", elapsed_ms);
    return 0;
}
